analysis/Resonator_fitting/fit_biased_lorentzian.m

```diff
@@ -0,0 +1,201 @@
+% Author/Date: Guilhem Ribeill, 
+%
+% Copyright 2016 Raytheon BBN Technologies
+
+function [fit_params, fit_vars] = fit_biased_lorentzian(freq, data, varargin)
+% FIT_BIASED_LORENTZIAN Fit frequency-amplitude data to a skewed
+% lorentizan. Returns parameters of fit and variance of parameters
+%
+%   [FIT_PARAMS, FIT_ERRORS] = FIT_BIASED_LORENTZIAN(FREQ, DATA) fits
+%   data to a biased lorentzian of the form
+%   f(x) = a / ((x - b)^2 + (c / 2)^2) + d*(x - b) + e
+%   For two-dimensional data fits a lorentzian to each slice of data whose 
+%   dimension matches that of frequency. For square matricies, a lorentzian
+%   is fit to each column of data. 
+%   Optional string-value parameters:
+%       - 'MaxIter': Maximum number of fminsearch iterations. Default 2000.
+%       - 'Plot': Plot result of fit
+persistent figHandles;
+if isempty(figHandles)
+        figHandles = struct();
+end
+parser = inputParser;
+defaultDim = 1;
+defaultMaxIter = 2000;
+addRequired(parser, 'freq', @isnumeric);
+addRequired(parser, 'data', @isnumeric);
+addOptional(parser, 'dims', defaultDim, @isnumeric);
+addParameter(parser, 'MaxIter', defaultMaxIter, @isnumeric);
+addParameter(parser, 'Plot', 0);
+parse(parser, freq, data, varargin{:});
+dims = parser.Results.dims;
+
+%validate inputs
+freq = freq(:);
+if isvector(data)
+    data = data(:);
+    N = length(data); M = 1;
+    if length(data) ~= length(freq)
+        error('Data and frequency dimensions do not match');
+    end
+elseif ~ismatrix(data)
+    error('Data has too many dimensions. I don''t know how to proceed.')
+else
+    [N, M] = size(data);
+    if N ~= length(freq) && M ~= length(freq)
+        error('Data dimensions do not match frequency vector dimensions.')
+    end
+    if strmatch('dims', parser.UsingDefaults)
+        if M == length(freq)
+            data = data';
+            t = M; M = N; N = t;
+        end
+    else
+        if dims == 2
+            data = data';
+            t = M; M = N; N = t;
+        end
+    end
+end
+
+options = optimset('MaxIter', parser.Results.MaxIter);
+
+fit_params = zeros(M, 5);
+fit_vars = zeros(M, 5);
+ssemin = zeros(M,1);
+
+for idx = 1:M  
+    p0 = initial_guess(freq, data(:,idx));
+    model = @(p)bias_lorentz_model(p, freq, data(:,idx));
+    [p, ssemin(idx)] = fminsearch(model, p0, options);
```

Comment:
Matt's guess is right.  `fminsearch` does okay even if it's not ideal, although someone should probably double check that I'm calculating confidence intervals correctly.